### DISCUSSION TITLE
Extraction efficiency as a function of NA

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2368] [Q&A]
Hello again friends from the MEEP community, 

This time I would like to ask for your guidance in order to be able to correctly evaluate the extraction efficiency from a dipole emitter source which is embedded in a flat semiconductor layer. I have taken as a reference for this implementation the tutorial: https://meep.readthedocs.io/en/latest/Python_Tutorials/Local_Density_of_States/ (Extraction Efficiency of a Light-Emitting Diode (LED)
). Specifically, I am just using the 2D implementation in cylindrical coordinates. 

In that tutorial, the extraction efficiency for a dipole emitter embedded in a planar semiconductor layer was evaluated as a function of the emitter's vertical offset with respect to the top surface interface.

Taking this as a starting point, I am interested in computing the extraction efficiency of a dipole emitter which is located at a fixed vertical position as a function of the "NA" of a collection objective which is meant be used for extracting the radiation propagating towards the vertical upper part of the simulation domain. 

In order to simulate this "objective" with a variable NA, I just simply defined a flux monitor surface along the x direction (a horizontal monitor) without including the vertical monitor that is used at the edge of the simulation domain. 

I then vary the extension of the flux monitor. By increasing the width of this surface I increase the NA of my fictitious collection objective. 

My simple intuition would tell me that as I increase the width of the flux monitor, I am able to collect more radiation from the emitter. Nevertheless, once I run the simulation for the just mentioned implementation I obtain the following curve:

![extraction_efficiency_meep_ref_detect_w_sweep_top_only](https://user-images.githubusercontent.com/49510503/211031242-34407353-725d-4a07-a931-30b59b810a0f.png)

You can see that the extraction efficiency first increases towards a maximum but suddenly starts to decrease, even if my collection monitor surface is larger. Above plot corresponds to the same material configuration used in the tutorial that I just mentioned.

As for my own material configuration (my dipole is embedded in GaAs, with a polymer material on top of GaAs and SiO2 below) I get the following curve:
![extraction_efficiency_res_80_ips_top_only_NA](https://user-images.githubusercontent.com/49510503/211031644-5c5e4056-d6cc-4e38-b12a-bcc64f711bbf.png)

Again, you can observe the same trend: the collection efficiency increases towards a max value from where then the obtained value starts to decrease. 

For the first configuration case the deviation from the maximum might not be too large (around 1%) but for my own configuration, you can see that the efficiency decreases by about 5%. 

Ideally, I would expect to get a curve similar to this:
![ext_eff_ref](https://user-images.githubusercontent.com/49510503/211031946-1b6b710f-0dc0-42dc-b617-d49690b07e16.PNG)

As you can observe, the extraction efficiency never decreases if the value of the collection NA is increased.

What could be the cause of the observed behavior?  Do you have any suggestions on what could be tried in order to verify what is going on? 

Again, thanks a lot for your help!

**Comment by oskooi** [comment]
A more efficient computational approach for obtaining extraction efficiency for all collection NAs using a *single* simulation is to use the [near-to-far field transformation feature](https://meep.readthedocs.io/en/latest/Python_User_Interface/#near-to-far-field-spectra) as demonstrated in [Tutorial/Radiation Pattern of an Antenna](https://meep.readthedocs.io/en/latest/Python_Tutorials/Near_to_Far_Field_Spectra/#radiation-pattern-of-an-antenna) for a dipole antenna in 2D Cartesian coordinates. See also [Tutorial/Focusing Properties of a Binary-Phase Zone Plate](https://meep.readthedocs.io/en/latest/Python_Tutorials/Cylindrical_Coordinates/#focusing-properties-of-a-binary-phase-zone-plate) which demonstrates this feature in cylindrical coordinates.

**Reply by kalosu** [reply]
Hi @oskooi , thanks a lot for your suggestion.

I have tried the following at the end:
1. I use a near to far field transformation for generating the far field components from my small simulation volume.
2. Using these far field components I compute the total power by evaluating the different Poynting flux components and integrating over the defined far field evaluation region.
3. For extracting the power from the source, I use the approach based on the evaluation of the LDOS. 

In order to validate this approach, I have used the test case of having a dipole emitter embedded in a homogeneous space sitting at the origin of the coordinate system. 

I started by first checking for the case of using a 2D simulation in cartesian coordinates: For the far field monitor I used a circle around the simulation domain with a radius equal to 1000 times the central simulation wavelength (same as in the antenna far field tutorial)

From the graph shown below you can see that as expected, all of the power emitted by the dipole is collected by the far field monitor, resulting in a total "extraction efficiency" equal to 100%.

![dipole_power_free_space_cart_coord_ext_eff](https://user-images.githubusercontent.com/49510503/213442549-00dcc6f9-7bb0-4198-82bc-2e6c1c2d3722.png)

In this case, I ran the simulation by increasing the number of sampling points used in the far field monitor. You can see that there is pretty much no difference in the results.

Then I moved to the cylindrical 2D case (this is what I am interested at the end). For this configuration, as you know the simulation domain is specified in cylindrical coordinates (r from 0 to the maximum value of my domain along r, phi has no extension and z goes from -max_z*0.5 to max_z*0.5). Again, since I wanted to simulate the case of having a dipole embedded in an open homogeneous space, I used a PML for both of the z sides and of course for the r side towards the end of the simulation domain.

In this case, I used the same radius for the far field monitor (1000 times the central wavelength) but the monitor is defined only for the "right half side" of the circle, i.e, only for points $x^{2} + z^{2 }= r^{2}$ for which x is equal or larger than 0. 

For this I am not plotting the ratio between the obtained powers but these two separate factors.

![dipole_power_free_space_cyl_coord_powers](https://user-images.githubusercontent.com/49510503/213448831-8619c8c2-8063-466f-ac40-f45a29d5cdf8.png)

The values shown in the plot are the following:
 0.0012637065053194476 for the power collected by the far field monitor and 4.029287646180604 for the total power emitted by the dipole.

As you can see, for the cylindrical coordinates implementation the result is not the expected one. 

Do you have any suggestion about what could I do in this case? If I am interested in modeling the extraction efficiency, could I also use the 2D simulation in cartesian coordinates? Would this be a valid approach? As far I as understand it, a 2D cylindrical simulation is equivalent to a 3D cartesian simulation for the case of having a rotationally symmetric material configuration, so I am not sure if I could use the results from the 2D cartesian simulations for "approximating" the results from such 3D system. 

I am also attaching the python script that I used for the cylindrical coordinates simulation, just in case you want to take a look to the code.

[dip_hom_ext_eff_cyl.zip](https://github.com/NanoComp/meep/files/10456841/dip_hom_ext_eff_cyl.zip)

**Comment by kalosu** [comment]
Hello again friends from the MEEP community, hello @oskooi 

In order to understand what is going on with the results from the cylindrical 2D simulation case, I have checked a couple of things more related to the energy flux.

Again, I started by using as a reference the case of a dipole in homogeneous space, sitting at the center of the coordinate system. I simulated this case first using 3D cartesian coordinates for which I included near2far surfaces around my dipole point. I then evaluated for the far field components along two different paths: circles formed by points in the far field (r equal to 1000/fcen) at the two transversal planes: xz and yz. In order to understand how is the energy flowing along different directions, I just plot first the values of the different poynting flux terms (px, py and pz) along the defined circles:

![dip_hom_space_3D_circle_fluxes_xz_yz_planes](https://user-images.githubusercontent.com/49510503/214005662-0213a402-acdc-4df2-9a47-26af6e21756d.png)

From above plot, it is clear how the energy mostly flows along the z and y directions for a dipole which is polarized along x, i.e, no energy is radiated along x for points found already in the far field. 

I then tried the same approach for the cylindrical 2D simulation case. You can see the results from this evaluation in the plot shown below:

first circle: within the rz plane, second circle: for points found at r=0, z=1000/fcen and phi from 0 to 2*pi

![dip_hom_space_cyl_flux_rz_phiz_planes_half_circ](https://user-images.githubusercontent.com/49510503/214006267-c2e992f1-be69-4e97-9f69-a4f34ad9c607.png)

Or through this plot: (first circle: again within the rz plane, second circle, for points found at r=1000/fcen, z=0 and phi from 0 to 2*pi)

![dip_hom_space_cyl_flux_rz_phiz_planes_full_circ](https://user-images.githubusercontent.com/49510503/214006673-70ee6fe3-be84-4597-bec9-16bbbf71950b.png)

From these two curves you can observe that there is a non zero energy flux along r, i.e, on the first plot you can observe how pr within the rz plane is non zero for an angle value equal to 0 (this would represent a point found at r=1000/fcen and z equal to 0). and from the second plot you can observe the same thing for pr along the phiz plane (again in this case the evaluation is done with r=100/fcen, z=0 and phi from 0 to 2*pi)

These results then make me ask the following questions:

1. For the cylindrical coordinate system case, is it expected that there exists a non zero energy flow along r into the far field for phi=0? Is it not the same case as for cartesian coordinates in where there is no energy flow along the dipole's oscillation direction? 
2. If I understand it correctly, a 2D cylindrical simulation case is equivalent to a 3D simulation case. If this is so, does this mean that I should consider a "2D plane" (points found in the rphi plane) for evaluating the extraction efficiency towards the upper side of the simulation domain? (even though the simulation is in 2D (rz plane) I should consider all points found on a 2D plane on top of the original 3D domain) 
3. If my interpretation on point 2 is correct, is there any way to take into account the phi dependence in an analytical manner? i.e, can I evaluate only for points found within the 2D original plane (rz plane) and then use some sort of scaling factor for taking into account the cylindrical symmetry of the problem? 

As always, thanks a lot for your help and feedback!
